### PR TITLE
Add NAS gateway support

### LINKED ISSUE #5062: Add support NAS Gateway

### PR DESCRIPTION
**Pull request by poornas**
Fixes #5062 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

File: cmd/gateway/nas/gateway-nas.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+// Production - nas gateway is not production ready.
+func (g *NAS) Production() bool {
+	return false
+}
+
+// nasObjects implements gateway for Minio and S3 compatible object storage servers.
+type nasObjects struct {
+	minio.GatewayUnsupported
+	objAPI minio.ObjectLayer
+}
```
Comment (krishnasrinivas) [inline cmd/gateway/nas/gateway-nas.go]: You can do this and remove all the other methods on nasObjects (minio.fsObjects should be exported as minio.FSObjects)
```
// nasObjects implements gateway for Minio and S3 compatible object storage servers.
type nasObjects struct {
        minio.FSObjects
}

func (l *nasObjects) IsNotificationSupported() bool {
        false
}

```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by krishnasrinivas** [comment]
@poornas travis is failing

**Comment by poornas** [comment]
@krishnasrinivas ,travis now passes

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

File: cmd/gateway/nas/gateway-nas.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+// NewGatewayLayer returns nas gatewaylayer.
+func (g *NAS) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error) {
+	var err error
+	newObject, err := minio.NewFSObjectLayer(g.host)
+	if err != nil {
+		return nil, err
+	}
+	return newObject, nil
```
Comment (krishnasrinivas) [inline cmd/gateway/nas/gateway-nas.go]: `nasObjects` is not being used here?

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#12ceb67d-654a-46d2-ba52-e1c369840064) |
<a name="12ceb67d-654a-46d2-ba52-e1c369840064"></a>
### 5516-25cd5a2/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31099
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 4ffabbccb889:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 1 seconds
PHP Fatal error:  Uncaught exception 'Aws\S3\Exception\S3Exception' with message 'Error executing "DeleteObject" on "http://minikube:31099/aws-sdk-php-5a834f73cd02b/obj1"; AWS HTTP error: Server error: `DELETE http://minikube:31099/aws-sdk-php-5a834f73cd02b/obj1` resulted in a `503 Service Unavailable` response:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, pl (truncated...)
 XMinioServerNotInitialized (server): Server not initialized, please try again. - <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/aws-sdk-php-5a834f73cd02b/obj1</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>'

GuzzleHttp\Exception\ServerException: Server error: `PUT http://minikube:31099/aws-sdk-php-5a834f73cd02b` resulted in a `503 Service Unavailable` response:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioSe in /mint/run/core/aws-sdk-php/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php on line 191

Executed 1 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#c352415a-ff8b-41c4-9d97-753eae4b55b8) |
<a name="c352415a-ff8b-41c4-9d97-753eae4b55b8"></a>
### 5516-45f6c0b/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30557
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 896a7152cfe1:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... done in 3 minutes and 48 seconds
(3/12) Running aws-sdk-ruby tests ... done in 59 seconds
(4/12) Running awscli tests ... FAILED in 1 minutes and 6 seconds
{
  "name": "awscli",
  "duration": 14916,
  "function": "aws --endpoint-url http://minikube:30557 s3api head-object --bucket \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. --key prefix/directory/\n",
  "status": "FAIL",
  "error": "usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]To see help text, you can run:  aws help  aws <command> help  aws <command> <subcommand> helpUnknown options: occurred, (BucketAlreadyOwnedByYou), when, calling, the, CreateBucket, operation:, Your, previous, request, to, create, the, named, bucket, succeeded, and, you, already, own, it., error"
}

Executed 3 out of 12 tests successfully.
```

File: docs/gateway/nas.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Minio NAS Gateway [![Slack](https://slack.minio.io/slack?type=svg)](https://slack.minio.io)
+Minio Gateway adds Amazon S3 compatibility to NAS storage.
+
+## Run Minio Gateway for NAS Storage
+
+### Using Docker
+```
+docker run -p 9000:9000 --name nas-s3 \
```
Comment (harshavardhana) [inline docs/gateway/nas.md]: We need to write a bit more documentation here can you check with @abperiasamy about this ?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#badb4a7d-43a4-4e8f-b1f9-9a1ebf89e0ea) |
<a name="badb4a7d-43a4-4e8f-b1f9-9a1ebf89e0ea"></a>
### 5516-c707ddf/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31988
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ec7e05305e75:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-p0p2mz990fy0qz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 417,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:31988/aws-sdk-go-test-p0p2mz990fy0qz: dial tcp 192.168.122.45:31988: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

File: docs/gateway/nas.md (hunk 3 above)
Comment (poornas) [reply docs/gateway/nas.md]: updated documentation after checking with @abperiasamy

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#3dbb7aef-16fc-42b9-bcdd-cbe4304ebb8e) |
| mint-dist-xl.sh | :x: [more...](#b42f3ef3-71bd-4380-8e52-df05a88d6396) |
<a name="3dbb7aef-16fc-42b9-bcdd-cbe4304ebb8e"></a>
### 5516-a0dfeb0/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31747
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 99f097a6a230:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 0 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 50 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8684cccc698"
  },
  "duration": "40175",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```
<a name="b42f3ef3-71bd-4380-8e52-df05a88d6396"></a>
### 5516-a0dfeb0/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30416
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ddd3b8574f3d:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-nzju29fyo5kucw",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 298,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30416/aws-sdk-go-test-nzju29fyo5kucw: dial tcp 192.168.122.45:30416: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

File: docs/gateway/nas.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# Minio NAS Gateway [![Slack](https://slack.minio.io/slack?type=svg)](https://slack.minio.io)
+Minio Gateway adds Amazon S3 compatibility to NAS storage. You may run multiple minio instances on the same shared NAS volume as a distributed object gateway.
+
+## Run Minio Gateway for NAS Storage
+
```
Comment (harshavardhana) [inline docs/gateway/nas.md]: remove this space.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#96bc6cd3-2455-4368-99a0-8a3e1cf24785) |
| mint-dist-xl.sh | :x: [more...](#1f87e339-b6cc-4c3a-9d16-e6d746ab6821) |
| mint-gateway-azure.sh | :x: [more...](#13004825-9d43-461b-b1f2-3afea7ba4ec2) |
<a name="13004825-9d43-461b-b1f2-3afea7ba4ec2"></a>
### 5516-3a3852a/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30578
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 5bc6905918ed:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... done in 3 minutes and 50 seconds
(3/12) Running aws-sdk-ruby tests ... done in 39 seconds
(4/12) Running awscli tests ... FAILED in 2 minutes and 17 seconds
{
  "name": "awscli",
  "duration": 7262,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 3 out of 12 tests successfully.
```
<a name="1f87e339-b6cc-4c3a-9d16-e6d746ab6821"></a>
### 5516-3a3852a/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30766
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 9138502322ea:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-um99mcw0lptez1",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 375,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30766/aws-sdk-go-test-um99mcw0lptez1: dial tcp 192.168.122.45:30766: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```
<a name="96bc6cd3-2455-4368-99a0-8a3e1cf24785"></a>
### 5516-3a3852a/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32060
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp f1c91892b5b1:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 0 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 47 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a868c7c57291"
  },
  "duration": "40190",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```

File: docs/gateway/nas.md (hunk 4 above)
Comment (poornas) [reply docs/gateway/nas.md]: done

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#3d09b7be-62dc-4802-b191-71b4dff06ab6) |
| mint-dist-xl.sh | :x: [more...](#8e0a7341-7d57-4e1b-9bf3-b906a37e0a35) |
<a name="3d09b7be-62dc-4802-b191-71b4dff06ab6"></a>
### 5516-b49975d/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30588
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 97f0fae47065:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 46 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a871cc07bea9"
  },
  "duration": "40104",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```
<a name="8e0a7341-7d57-4e1b-9bf3-b906a37e0a35"></a>
### 5516-b49975d/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30836
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 9ab9bb4169b4:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-1sczcxmgo9cf4y",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 313,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30836/aws-sdk-go-test-1sczcxmgo9cf4y: dial tcp 192.168.122.45:30836: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Review by krishnasrinivas — COMMENTED** [review]
@poornas you'll need to remove `buildcoveragecoverage.txt`

You can add buildcoveragecoverage.txt to .gitignore

File: cmd/gateway/nas/gateway-nas.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	var err error
+	newObject, err := minio.NewFSObjectLayer(g.host)
+	if err != nil {
+		return nil, err
+	}
+	return &nasObjects{newObject.(*minio.FSObjects)}, nil
+}
+
+// Production - nas gateway is not production ready.
+func (g *NAS) Production() bool {
```
Comment (krishnasrinivas) [inline cmd/gateway/nas/gateway-nas.go]: NAS gateway is actually production ready

Comment (poornas) [reply cmd/gateway/nas/gateway-nas.go]: done.

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#d94dcba8-13a9-42d5-b938-1fee229a833b) |
| mint-dist-xl.sh | :x: [more...](#19b8fda2-c3e2-4590-adfa-7ec49133c584) |
<a name="19b8fda2-c3e2-4590-adfa-7ec49133c584"></a>
### 5516-6800121/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30983
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp a1dc1fb7d2ca:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-0fkue5shpj1l2t",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 4524,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30983/aws-sdk-go-test-0fkue5shpj1l2t: dial tcp 192.168.122.45:30983: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```
<a name="d94dcba8-13a9-42d5-b938-1fee229a833b"></a>
### 5516-6800121/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30564
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 97e103dac6ca:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 47 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8c652dad36d"
  },
  "duration": "40081",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**